pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,166 @@
+
+    @property
+    def filename(self):
+        return self.path_parts[3]
+
+    @property
+    def store_qs(self):
+        qs = Store.objects
+        if self.lang_code:
+            qs = qs.filter(translation_project__language__code=self.lang_code)
+        if self.proj_code:
+            qs = qs.filter(translation_project__project__code=self.proj_code)
+        if self.use_regex:
+            if self.lang_code or self.proj_code:
+                qs = qs.filter(pootle_path__regex="^%s" % self.path)
+        else:
+            qs = qs.filter(pootle_path__startswith=self.path)
+        return qs
+
+    @property
+    def use_regex(self):
+        return (
+            not self.lang_code
+            or not self.proj_code)
+
+    @property
+    def path_regex(self):
+        regex = "/[^/]*"
+        return self.vf.location.replace(
+            "/{LANG}", regex).replace("/{PROJ}", regex)
+
+    @property
+    def regexes(self):
+        return (
+            "%s%s" % (self.path_regex, rule)
+            for rule
+            in self.filter_rules)
+
+    @property
+    def path(self):
+        return (
+            self.path_regex
+            if self.use_regex
+            else self.vf.location)
+
+    def get_rule_path(self, rule):
+        return "%s%s" % (self.path, rule)
+
+    def filter_by_rule(self, qs, rule):
+        rule_path = self.get_rule_path(rule)
+        is_file = "." in os.path.basename(rule)
+        if self.use_regex:
+            return (
+                qs.filter(pootle_path__regex="^%s$" % rule_path)
+                if is_file
+                else qs.filter(pootle_path__regex="^%s" % rule_path))
+        return (
+            qs.filter(pootle_path=rule_path)
+            if is_file
+            else qs.filter(pootle_path__startswith=rule_path))
+
+    def filter_from_rules(self, qs):
+        if not self.filter_rules:
+            return qs
+        filtered_qs = qs.none()
+        for rule in self.filter_rules:
+            filtered_qs = filtered_qs | self.filter_by_rule(qs, rule)
+        return filtered_qs
+
+    def find_matching_stores(self):
+        return self.filter_from_rules(self.store_qs)
+
+    def path_matches(self, path):
+        """Returns bool of whether path is valid for this VF.
+        """
+        (lang_code, proj_code,
+         dir_path, filename) = split_pootle_path(path)
+        dir_path = "/".join(dir_path.split("/")[1:])
+        no_match = (
+            self.lang_code and lang_code != self.lang_code
+            or (self.proj_code and proj_code != self.proj_code)
+            or (self.dir_path and not self.dir_path.startswith(dir_path)))
+        if no_match:
+            return False
+        # the path matches the vfolder - but we need to check the filter rules
+        # to ensure any remaining dir_path also matches
+        remaining_dir_path = dir_path[len(self.dir_path):]
+        for rule in self.filter_rules:
+            if rule.startswith(remaining_dir_path):
```

Comment:
You are assuming that vfolder location can only be `/lang/proj/` but it can actually can exists a vfolder with location `/af/firefox/browser/chrome/browser/preferences/` and rules `advanced.dtd.po,colors.dtd.po`. I suggest you to expand location+rules to get all possible combinations and then check if passed path matches or starts with any of those.